### PR TITLE
test: deflake TSAN test failure with giant payload in multiplex_integration_test

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: some tests with giant request/response payload in multiplex_integration_tests.cc have been flaky at FakeHttpConnection::waitForNewStream() under TSAN because of the contention between the main test thread and the fake upstream thread. This contention exists in all tests that uses non-autonomous fake upstream, but more severe in tests with large payload because their clients have more IO events to handle while holding the mutex.

This change modifies two things to mitigate the synchronization cost:
1. Change large payload tests to use autonomous fake upstream, so that the synchronization between the main test thread and the fake upstream thread can be eliminated.
2. Adjust the timeout of waitForWithDispatcherRun() to respect TSAN_TIMEOUT_FACTOR, so that the main test thread doesn't too aggressively acquire the mutex.

Risk Level: low, test only
Testing: existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by jmarantz** [comment]
This OK but WDYT of trying to use simulated time for the flaky tests? Is there some reason sim-time can't work for those?

@botengyao FYI

**Comment by danzh2010** [comment]
> This OK but WDYT of trying to use simulated time for the flaky tests? Is there some reason sim-time can't work for those?

I'm not aware of how simulated time system could help with contention in integration tests. Is there any example of it?

**Comment by jmarantz** [comment]
I have a question about your question :) Is contention the reason that the test flakes? Why would that be? I can understand why contention results in tests being slow, but not why contention would cause tests to flake.

Most often when I see someone change a time constant to try to resolve test flakes, think that's because the real-time delays when running the test are hard to predict and you are trying to tweak constants to make it less likely that a flake occurs due to time-constant-ordered events happening in an unexpected order.

SimulatedTimeSystem attacks that directly by taking direct control of time from the test harness, and enabling deterministic ordering of timer callbacks, at least in scenarios where the test harness advances time in increments that make it unambiguous which timers are runnable.

There are currently a large number of unit tests using simulated time, but only a handful of integration tests using it (`cd test/integration; grep Simulated *.cc *.h`) to see which ones. But I think the ones that are flaky are the ones we should look to shift to simulated time first.

Having said that, as of this writing, gRPC presents challenges for simulated time since it doesn't have a mechanism to inject it (yet). For the simulated-time system to work properly, all code that is reading the current time needs to be going through the TimeSource interface, and not bypassing it by going directly to std::chrono, absl Time, or the low level time primitives.

**Comment by danzh2010** [comment]
> I have a question about your question :) Is contention the reason that the test flakes? Why would that be? I can understand why contention results in tests being slow, but not why contention would cause tests to flake.

The flakiness is in TSAN tests where the main test thread waitFoo(timeout) methods keep acquiring a mutex while waiting for the condition Foo, which can only be met if the fake upstream acquires the same mutex and call setFoo(). And TSAN excessively starves setFoo() still the waitFoo() reaches its timeout...

> Most often when I see someone change a time constant to try to resolve test flakes, think that's because the real-time delays when running the test are hard to predict and you are trying to tweak constants to make it less likely that a flake occurs due to time-constant-ordered events happening in an unexpected order.
> 
> SimulatedTimeSystem attacks that directly by taking direct control of time from the test harness, and enabling deterministic ordering of timer callbacks, at least in scenarios where the test harness advances time in increments that make it unambiguous which timers are runnable.

From your description about SimulatedTimeSystem, it seems to tackle a different issue rather than contention though. And those giant request/response e2e tests are meant to test the flow control interaction between client-Envoy-upstream as close to production as possible, so I'm not sure if we want to introduce a simulated time.

IMO the mutex to check fake upstream status during proxying a request and response isn't really necessary in most cases, as long as the response itself makes sense.

File: test/integration/http_integration.cc [hunk 1]
```diff
@@ -591,6 +591,43 @@ void HttpIntegrationTest::testRouterRequestAndResponseWithBody(
   checkSimpleRequestSuccess(request_size, response_size, response.get());
 }
 
+void HttpIntegrationTest::testGiantRequestAndResponse(uint64_t request_size, uint64_t response_size,
+                                                      bool set_content_length_header,
+                                                      std::chrono::milliseconds timeout) {
+  autonomous_upstream_ = true;
+#if defined(ENVOY_CONFIG_COVERAGE)
```
Comment (jmarantz) [inline test/integration/http_integration.cc:598]: nit `#ifdef ENVOY_CONFIG_COVERAGE` and comment why this is helpful.

File: test/integration/http_integration.cc [hunk 2]
```diff
@@ -591,6 +591,43 @@ void HttpIntegrationTest::testRouterRequestAndResponseWithBody(
...
+  // https://github.com/envoyproxy/envoy/issues/19595
+  ENVOY_LOG_MISC(warn, "manually lowering logs to error");
```
Comment (jmarantz) [inline test/integration/http_integration.cc:600]: I'm not sure what `lowering` means here; maybe s/lowering/setting/ ?

File: test/integration/multiplexed_integration_test.cc [hunk 3]
```diff
@@ -57,23 +57,20 @@ TEST_P(MultiplexedIntegrationTest, RouterRequestAndResponseWithGiantBodyNoBuffer
   ENVOY_LOG_MISC(warn, "manually lowering logs to error");
   LogLevelSetter save_levels(spdlog::level::err);
   config_helper_.addConfigModifier(ConfigHelper::adjustUpstreamTimeoutForTsan);
-  testRouterRequestAndResponseWithBody(10 * 1024 * 1024, 10 * 1024 * 1024, false, false, nullptr,
-                                       TSAN_TIMEOUT_FACTOR * TestUtility::DefaultTimeout);
+  testGiantRequestAndResponse(10 * 1024 * 1024, 10 * 1024 * 1024, false);
```
Comment (jmarantz) [inline test/integration/multiplexed_integration_test.cc:60]: nit: make const for this 10M value

**Review by jmarantz — COMMENTED** [review]
I think the changes you have to improve the mutexing are independent of timing flakes, so lgtm for that. I agree that's not related to taking control of time.

RE sim-time vs real-time; I think this transition would be better in the long run. Ideally I think close to 100% of tests should be sim-time and that should be our default mechanism for creating new tests. But can you leave a TODO for now?

**Comment by danzh2010** [comment]
> RE sim-time vs real-time; I think this transition would be better in the long run. Ideally I think close to 100% of tests should be sim-time and that should be our default mechanism for creating new tests. But can you leave a TODO for now?

done

**Review by jmarantz — APPROVED** [review]
/assign-from @envoyproxy/senior-maintainers

**Comment by jmarantz** [comment]
Thanks!

**Review by zuercher — APPROVED** [review]

**Pull request merged**